source/common/network/filter_manager_impl.h

```diff
@@ -11,34 +12,109 @@ namespace Envoy {
 namespace Network {
 
 /**
- * Interface used to obtain read/write buffers.
+ * Declaration of StreamBuffer backwards compatible with the former BufferSource::StreamBuffer.
  */
-class BufferSource {
+class StreamBufferSource {
 public:
-  virtual ~BufferSource() {}
+  virtual ~StreamBufferSource() {}
 
   struct StreamBuffer {
     Buffer::Instance& buffer;
     bool end_stream;
   };
+};
+
+/**
+ * Interface used to obtain read buffers.
+ */
+class ReadBufferSource : public virtual StreamBufferSource {
+public:
+  virtual ~ReadBufferSource() {}
 
   /**
    * Fetch the read buffer for the source.
    */
   virtual StreamBuffer getReadBuffer() PURE;
+};
+
+/**
+ * Interface used to obtain write buffers.
+ */
+class WriteBufferSource : public virtual StreamBufferSource {
+public:
+  virtual ~WriteBufferSource() {}
 
   /**
    * Fetch the write buffer for the source.
    */
   virtual StreamBuffer getWriteBuffer() PURE;
 };
 
+/**
+ * Interface used to obtain read/write buffers.
+ */
+class BufferSource : public ReadBufferSource, public WriteBufferSource {
```

Comment:
Is this needed or can users just derive from both if needed?

Reply:
I removed `BufferSource` and made `FilterManagerConnection` to extend `ReadBufferSource` and `WriteBufferSource`